### ISSUE TITLE
separate root-srv go code from html and make it a library

### ISSUE THREAD

**Issue by vkuznet** [go-hep/hep#389]
I'm interested in separation of root-srv from html (presentation) layer. The later can be application specific and use different html/css/js tools which will allow to embed root-srv into other web applications which by itself should be possible if root-srv will be served as package/library.

**Comment by sbinet** [comment]
what are the parts that you'd like to be able to reuse? and at what level?

- do you want to be able to create a `root-srv/server.Server` application, change the default html page and call it a day?
- create a `server` value and add new end points?
- just be able to create end points in your own application and associate them with, say, `server.PlotH1`, `server.PlotH2`, ... ?

I could imagine the following API:

```go
package srrv // import "go-hep.org/x/hep/groot/rsrv"

// PlotH1 returns a JSON payload of the 1-dim histogram extracted from
// the query in the request:
//  - file: local/remote path to the ROOT file containing the 1-dim histogram
//          to be plotted
//  - path: path to the 1-dim histogram to be plotted
//  - name: name of the 1-dim histogram to be plotted
func (*Server) PlotH1(w http.ResponseWriter, r *http.Request) { ... }
```

**Comment by vkuznet** [comment]
Sebastien,
I think we need option #3 as shown in your example. The server part is trivial
in Go and can be different in apps, e.g. someone can run http or https, use
different auth layer, etc. The frontend JS libraries may be different
among projects as well. Therefore, we need bare APIs to serve plots
and/or stats about given list of files.
Thanks,
Valentin.

**Comment by sbinet** [comment]
ok, I'll start something along these lines.

**Comment by sbinet** [comment]
see:
- https://github.com/go-hep/hep/pull/395

still hugely WIP.

additional ideas/questions that popped up during this first stab:
- gRPC?
- interoperability with other (web) display backends? (_e.g._ https://github.com/HSF/Visualization)
- JSON representation of histograms for ROOT, YODA, ReAna, HepData, etc...

**Comment by vkuznet** [comment]
Sebastien,
in terms of additions. I would vote for JSON representation support first.
The gRPC may not be required but in a future may pop-up. And, support for
other display bakcends is not yet a table. Let's get things working in bare
metal.

Do you mind to add all docs to public functions and structs that they will appear on godoc?

Thanks,
Valentin

**Comment by sbinet** [comment]
Of course.
It was really just a first stab to get a very first rough idea of where I was going :)

I'll properly document all of this (especially the upload/open distinction and call order.)

**Comment by sbinet** [comment]
closed by #395
